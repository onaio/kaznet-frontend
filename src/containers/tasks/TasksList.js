// Smart component that renders the Task list view
import React, { Component } from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import { bindActionCreators } from "redux";
import Moment from "react-moment";
import queryString from "query-string";
import { Badge } from "reactstrap";

import * as taskActions from "../../store/tasks/actions";
import * as globalActions from "../../store/global/actions";
import * as taskSelectors from "../../store/tasks/reducer";
import * as globalSelectors from "../../store/global/reducer";
import * as constants from "../../constants.js";

import ListView from "../../components/ListView";
import ElementMap from "../ElementMap";

import "./TaskList.css";

export class TasksList extends Component {
  async componentDidMount() {
    this.props.showListTitle();
    this.props.changePageTitle("Tasks");
    this.props.changePageTitleButton("+ Create Task");
    this.props.changePageTarget("/tasks/new");

    let { search } = queryString.parse(this.props.location.search);
    const { page } = queryString.parse(this.props.location.search);

    if (search === undefined) {
      search = "";
    }
    this.props.searchVal(search);
    let pageNumber = Number(page);

    if (isNaN(pageNumber)) {
      pageNumber = 1;
    }

    if (search === "" && pageNumber) {
      await this.props.fetchTasks(
        `${constants.API_ENDPOINT}/tasks/?search=${search}&page=${pageNumber}`
      );
      this.props.changePageNumber(pageNumber);
    } else if (search !== "" && pageNumber) {
      await this.props.fetchTasks(
        `${constants.API_ENDPOINT}/tasks/?search=${search}&page=${pageNumber}`
      );
      this.props.changePageNumber(pageNumber);
    } else {
      this.props.fetchTasks();
    }
  }

  componentDidUpdate(prevProps) {
    let { search } = queryString.parse(this.props.location.search);
    if (search === undefined) {
      search = "";
    }
    const { page } = queryString.parse(this.props.location.search);
    if (Number(page) !== this.props.currentPage && !isNaN(page)) {
      const pageNumber = Number(page);
      this.props.fetchTasks(
        `${constants.API_ENDPOINT}/tasks/?search=${search}&page=${pageNumber}`
      );
      this.props.changePageNumber(pageNumber);
    } else {
      this.props.fetchTasks();
    }
  }

  render() {
    if (!this.props.rowsById) return this.renderLoading();
    return (
      <div className="TasksList">
        <ListView
          renderHeaders={this.renderHeaders}
          rowsIdArray={this.props.rowsIdArray}
          rowsById={this.props.rowsById}
          renderRow={this.renderRow}
          endpoint={"tasks"}
          pageLinks={this.props.pageLinks}
          totalPages={this.props.totalPages}
          currentPage={this.props.currentPage}
          firstPage={this.props.firstPage}
          lastPage={this.props.lastPage}
          searchVal={this.props.searchParam}
        />
      </div>
    );
  }

  renderLoading() {
    return <p>Loading...</p>;
  }

  renderHeaders() {
    const headerItems = [
      "Status",
      "Name",
      "Need Review",
      "Created",
      "Expires",
      "Form"
    ];
    return <ElementMap items={headerItems} HTMLTag="th" />;
  }

  renderRow(row) {
    const rowItems = [
      <Badge
        key={row.id}
        className={"task-badge status-" + row.attributes.status}
      >
        {row.attributes.status_display}
      </Badge>,
      <Link to={`/tasks/${row.id}`} key="link_to">
        {row.attributes.name}
      </Link>,
      row.attributes.pending_submissions_count +
        "/" +
        row.attributes.submission_count,
      <Moment key={row.id} format="DD-MM-YYYY">
        {row.attributes.start}
      </Moment>,
      <Moment key={row.id} format="DD-MM-YYYY">
        {row.attributes.end}
      </Moment>,
      row.attributes.xform_title
    ];
    return <ElementMap items={rowItems} HTMLTag="td" />;
  }
}

// which props do we want to inject, given the global store state?
function mapStateToProps(state) {
  return {
    rowsById: taskSelectors.getTasksById(state),
    rowsIdArray: taskSelectors.getTasksIdArray(state),
    totalPages: taskSelectors.getTotalPages(state),
    currentPage: taskSelectors.getCurrentPage(state),
    pageLinks: taskSelectors.getPageLinks(state),
    firstPage: taskSelectors.getFirstPage(state),
    lastPage: taskSelectors.getLastPage(state),
    searchParam: globalSelectors.getSearchValue(state)
  };
}

function mapDispatchToProps(dispatch) {
  return bindActionCreators(
    {
      fetchTasks: taskActions.fetchTasks,
      changePageNumber: taskActions.changePageNumber,
      changePageTitle: globalActions.changePageTitle,
      changePageTitleButton: globalActions.changePageTitleButton,
      changePageTarget: globalActions.changePageTarget,
      showListTitle: globalActions.toggleDetailTitleOff,
      searchVal: globalActions.getSearchVal
    },
    dispatch
  );
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(TasksList);
