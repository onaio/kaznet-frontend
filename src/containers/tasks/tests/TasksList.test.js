// Test TasksList
import React from "react";
import { shallow, mount } from "enzyme";
import toJson from "enzyme-to-json";
import { Router } from "react-router";
import createBrowserHistory from "history/createBrowserHistory";

import { TasksList } from "../TasksList";
import * as fixtures from "../../../store/tasks/tests/fixtures";

const history = createBrowserHistory();

describe("containers/task/TasksList", () => {
  it("renders without crashing", () => {
    shallow(
      <TasksList
        fetchTasks={function() {}}
        changePageNumber={function() {}}
        changePageTitle={function() {}}
        changePageTitleButton={function() {}}
        changePageTarget={function() {}}
        showListTitle={function() {}}
      />
    );
  });

  it("renders task list correctly", () => {
    const wrapper = mount(
      <Router history={history}>
        <TasksList
          fetchTasks={function() {}}
          changePageNumber={function() {}}
          changePageTitle={function() {}}
          changePageTitleButton={function() {}}
          changePageTarget={function() {}}
          showListTitle={function() {}}
          rowsById={fixtures.tasksById}
          rowsIdArray={fixtures.tasksIdArray}
          endpoint={"tasks"}
          pageLinks={fixtures.pageLinks}
          totalPages={fixtures.totalPages}
          currentPage={fixtures.currentPage}
        />
      </Router>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.unmount();
  });
});
