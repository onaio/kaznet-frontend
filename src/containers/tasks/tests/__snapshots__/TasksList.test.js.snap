// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/task/TasksList renders task list correctly 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <TasksList
    changePageNumber={[Function]}
    changePageTarget={[Function]}
    changePageTitle={[Function]}
    changePageTitleButton={[Function]}
    currentPage={1}
    endpoint="tasks"
    fetchTasks={[Function]}
    pageLinks={
      Object {
        "first": "http://localhost:8000/api/v1/tasks/?page=1",
        "last": "http://localhost:8000/api/v1/tasks/?page=1",
        "next": null,
        "prev": null,
      }
    }
    rowsById={
      Object {
        "1": Object {
          "attributes": Object {
            "approved_submissions_count": 1,
            "created": "2018-06-21T12:15:45.349936+03:00",
            "current_bounty_amount": "55.00 KES",
            "description": "This is pure joy.",
            "end": "2019-01-30T12:17:40+03:00",
            "estimated_time": "4 01:15:20",
            "modified": "2018-06-21T12:22:09.703447+03:00",
            "name": "Awesome Task",
            "pending_submissions_count": 0,
            "rejected_submissions_count": 0,
            "required_expertise": "1",
            "required_expertise_display": "Beginner",
            "start": "2018-06-21T12:15:45+03:00",
            "status": "a",
            "status_display": "Active",
            "submission_count": 1,
            "target_id": 1,
            "timing_rule": "RRULE:FREQ=DAILY;INTERVAL=10;COUNT=12",
            "total_bounty_payout": "55.00 KES",
            "total_submission_target": 100,
            "user_submission_target": 100,
            "xform_title": "Form 1",
          },
          "id": "1",
          "relationships": Object {
            "bounty": Object {
              "data": Object {
                "id": "2",
                "type": "Bounty",
              },
            },
            "client": Object {
              "data": Object {
                "id": "7",
                "type": "Client",
              },
            },
            "locations": Object {
              "data": Array [
                Object {
                  "id": "1",
                  "type": "Location",
                },
              ],
              "meta": Object {
                "count": 1,
              },
            },
            "parent": Object {
              "data": null,
            },
            "segment_rules": Object {
              "data": Array [],
              "meta": Object {
                "count": 0,
              },
            },
            "target_content_type": Object {
              "data": Object {
                "id": "16",
                "type": "ContentType",
              },
            },
          },
          "type": "Task",
        },
        "2": Object {
          "attributes": Object {
            "approved_submissions_count": 2,
            "created": "2018-06-21T12:15:45.363700+03:00",
            "current_bounty_amount": "150000.00 KES",
            "description": "The best and most delicious task ever.",
            "end": null,
            "estimated_time": "00:00:19",
            "modified": "2018-06-21T12:25:06.164454+03:00",
            "name": "Kill Bill",
            "pending_submissions_count": 0,
            "rejected_submissions_count": 0,
            "required_expertise": "3",
            "required_expertise_display": "Advanced",
            "start": "2018-08-17T12:15:45+03:00",
            "status": "s",
            "status_display": "Scheduled",
            "submission_count": 2,
            "target_id": 2,
            "timing_rule": "RRULE:FREQ=DAILY;INTERVAL=1;COUNT=500",
            "total_bounty_payout": "300000.00 KES",
            "total_submission_target": 200,
            "user_submission_target": 2,
            "xform_title": "Form 2",
          },
          "id": "2",
          "relationships": Object {
            "bounty": Object {
              "data": Object {
                "id": "1",
                "type": "Bounty",
              },
            },
            "client": Object {
              "data": Object {
                "id": "2",
                "type": "Client",
              },
            },
            "locations": Object {
              "data": Array [
                Object {
                  "id": "2",
                  "type": "Location",
                },
              ],
              "meta": Object {
                "count": 1,
              },
            },
            "parent": Object {
              "data": null,
            },
            "segment_rules": Object {
              "data": Array [],
              "meta": Object {
                "count": 0,
              },
            },
            "target_content_type": Object {
              "data": Object {
                "id": "16",
                "type": "ContentType",
              },
            },
          },
          "type": "Task",
        },
      }
    }
    rowsIdArray={
      Array [
        "1",
        "2",
      ]
    }
    showListTitle={[Function]}
  >
    <div
      className="TasksList"
    >
      <ListView
        currentPage={1}
        endpoint="tasks"
        pageLinks={
          Object {
            "first": "http://localhost:8000/api/v1/tasks/?page=1",
            "last": "http://localhost:8000/api/v1/tasks/?page=1",
            "next": null,
            "prev": null,
          }
        }
        renderHeaders={[Function]}
        renderRow={[Function]}
        rowsById={
          Object {
            "1": Object {
              "attributes": Object {
                "approved_submissions_count": 1,
                "created": "2018-06-21T12:15:45.349936+03:00",
                "current_bounty_amount": "55.00 KES",
                "description": "This is pure joy.",
                "end": "2019-01-30T12:17:40+03:00",
                "estimated_time": "4 01:15:20",
                "modified": "2018-06-21T12:22:09.703447+03:00",
                "name": "Awesome Task",
                "pending_submissions_count": 0,
                "rejected_submissions_count": 0,
                "required_expertise": "1",
                "required_expertise_display": "Beginner",
                "start": "2018-06-21T12:15:45+03:00",
                "status": "a",
                "status_display": "Active",
                "submission_count": 1,
                "target_id": 1,
                "timing_rule": "RRULE:FREQ=DAILY;INTERVAL=10;COUNT=12",
                "total_bounty_payout": "55.00 KES",
                "total_submission_target": 100,
                "user_submission_target": 100,
                "xform_title": "Form 1",
              },
              "id": "1",
              "relationships": Object {
                "bounty": Object {
                  "data": Object {
                    "id": "2",
                    "type": "Bounty",
                  },
                },
                "client": Object {
                  "data": Object {
                    "id": "7",
                    "type": "Client",
                  },
                },
                "locations": Object {
                  "data": Array [
                    Object {
                      "id": "1",
                      "type": "Location",
                    },
                  ],
                  "meta": Object {
                    "count": 1,
                  },
                },
                "parent": Object {
                  "data": null,
                },
                "segment_rules": Object {
                  "data": Array [],
                  "meta": Object {
                    "count": 0,
                  },
                },
                "target_content_type": Object {
                  "data": Object {
                    "id": "16",
                    "type": "ContentType",
                  },
                },
              },
              "type": "Task",
            },
            "2": Object {
              "attributes": Object {
                "approved_submissions_count": 2,
                "created": "2018-06-21T12:15:45.363700+03:00",
                "current_bounty_amount": "150000.00 KES",
                "description": "The best and most delicious task ever.",
                "end": null,
                "estimated_time": "00:00:19",
                "modified": "2018-06-21T12:25:06.164454+03:00",
                "name": "Kill Bill",
                "pending_submissions_count": 0,
                "rejected_submissions_count": 0,
                "required_expertise": "3",
                "required_expertise_display": "Advanced",
                "start": "2018-08-17T12:15:45+03:00",
                "status": "s",
                "status_display": "Scheduled",
                "submission_count": 2,
                "target_id": 2,
                "timing_rule": "RRULE:FREQ=DAILY;INTERVAL=1;COUNT=500",
                "total_bounty_payout": "300000.00 KES",
                "total_submission_target": 200,
                "user_submission_target": 2,
                "xform_title": "Form 2",
              },
              "id": "2",
              "relationships": Object {
                "bounty": Object {
                  "data": Object {
                    "id": "1",
                    "type": "Bounty",
                  },
                },
                "client": Object {
                  "data": Object {
                    "id": "2",
                    "type": "Client",
                  },
                },
                "locations": Object {
                  "data": Array [
                    Object {
                      "id": "2",
                      "type": "Location",
                    },
                  ],
                  "meta": Object {
                    "count": 1,
                  },
                },
                "parent": Object {
                  "data": null,
                },
                "segment_rules": Object {
                  "data": Array [],
                  "meta": Object {
                    "count": 0,
                  },
                },
                "target_content_type": Object {
                  "data": Object {
                    "id": "16",
                    "type": "ContentType",
                  },
                },
              },
              "type": "Task",
            },
          }
        }
        rowsIdArray={
          Array [
            "1",
            "2",
          ]
        }
      >
        <div>
          <Table
            bordered={true}
            className="kaznet-table"
            responsiveTag="div"
            tag="table"
          >
            <table
              className="kaznet-table table table-bordered"
            >
              <thead>
                <tr>
                  <ElementMap
                    HTMLTag="th"
                    items={
                      Array [
                        "Status",
                        "Name",
                        "Need Review",
                        "Created",
                        "Expires",
                        "Form",
                      ]
                    }
                  >
                    <th
                      key="0"
                    >
                      Status
                    </th>
                    <th
                      key="1"
                    >
                      Name
                    </th>
                    <th
                      key="2"
                    >
                      Need Review
                    </th>
                    <th
                      key="3"
                    >
                      Created
                    </th>
                    <th
                      key="4"
                    >
                      Expires
                    </th>
                    <th
                      key="5"
                    >
                      Form
                    </th>
                  </ElementMap>
                </tr>
              </thead>
              <tbody>
                <tr
                  key="1"
                >
                  <ElementMap
                    HTMLTag="td"
                    items={
                      Array [
                        "Active",
                        <Link
                          replace={false}
                          to="/tasks/1"
                      >
                          Awesome Task
                      </Link>,
                        "0/1",
                        <t
                          ago={false}
                          calendar={false}
                          decimal={false}
                          element="time"
                          filter={[Function]}
                          format="DD-MM-YYYY"
                          fromNow={false}
                          interval={60000}
                          onChange={[Function]}
                          toNow={false}
                          unit={null}
                          unix={false}
                          utc={false}
                      >
                          2018-06-21T12:15:45+03:00
                      </t>,
                        <t
                          ago={false}
                          calendar={false}
                          decimal={false}
                          element="time"
                          filter={[Function]}
                          format="DD-MM-YYYY"
                          fromNow={false}
                          interval={60000}
                          onChange={[Function]}
                          toNow={false}
                          unit={null}
                          unix={false}
                          utc={false}
                      >
                          2019-01-30T12:17:40+03:00
                      </t>,
                        "Form 1",
                      ]
                    }
                  >
                    <td
                      key="0"
                    >
                      Active
                    </td>
                    <td
                      key="1"
                    >
                      <Link
                        key="link_to"
                        replace={false}
                        to="/tasks/1"
                      >
                        <a
                          href="/tasks/1"
                          onClick={[Function]}
                        >
                          Awesome Task
                        </a>
                      </Link>
                    </td>
                    <td
                      key="2"
                    >
                      0/1
                    </td>
                    <td
                      key="3"
                    >
                      <t
                        ago={false}
                        calendar={false}
                        decimal={false}
                        element="time"
                        filter={[Function]}
                        format="DD-MM-YYYY"
                        fromNow={false}
                        interval={60000}
                        key="1"
                        onChange={[Function]}
                        toNow={false}
                        unit={null}
                        unix={false}
                        utc={false}
                      >
                        <time
                          dateTime={"2018-06-21T09:15:45.000Z"}
                        >
                          21-06-2018
                        </time>
                      </t>
                    </td>
                    <td
                      key="4"
                    >
                      <t
                        ago={false}
                        calendar={false}
                        decimal={false}
                        element="time"
                        filter={[Function]}
                        format="DD-MM-YYYY"
                        fromNow={false}
                        interval={60000}
                        key="1"
                        onChange={[Function]}
                        toNow={false}
                        unit={null}
                        unix={false}
                        utc={false}
                      >
                        <time
                          dateTime={"2019-01-30T09:17:40.000Z"}
                        >
                          30-01-2019
                        </time>
                      </t>
                    </td>
                    <td
                      key="5"
                    >
                      Form 1
                    </td>
                  </ElementMap>
                </tr>
                <tr
                  key="2"
                >
                  <ElementMap
                    HTMLTag="td"
                    items={
                      Array [
                        "Scheduled",
                        <Link
                          replace={false}
                          to="/tasks/2"
                      >
                          Kill Bill
                      </Link>,
                        "0/2",
                        <t
                          ago={false}
                          calendar={false}
                          decimal={false}
                          element="time"
                          filter={[Function]}
                          format="DD-MM-YYYY"
                          fromNow={false}
                          interval={60000}
                          onChange={[Function]}
                          toNow={false}
                          unit={null}
                          unix={false}
                          utc={false}
                      >
                          2018-08-17T12:15:45+03:00
                      </t>,
                        <t
                          ago={false}
                          calendar={false}
                          decimal={false}
                          element="time"
                          filter={[Function]}
                          format="DD-MM-YYYY"
                          fromNow={false}
                          interval={60000}
                          onChange={[Function]}
                          toNow={false}
                          unit={null}
                          unix={false}
                          utc={false}
                      />,
                        "Form 2",
                      ]
                    }
                  >
                    <td
                      key="0"
                    >
                      Scheduled
                    </td>
                    <td
                      key="1"
                    >
                      <Link
                        key="link_to"
                        replace={false}
                        to="/tasks/2"
                      >
                        <a
                          href="/tasks/2"
                          onClick={[Function]}
                        >
                          Kill Bill
                        </a>
                      </Link>
                    </td>
                    <td
                      key="2"
                    >
                      0/2
                    </td>
                    <td
                      key="3"
                    >
                      <t
                        ago={false}
                        calendar={false}
                        decimal={false}
                        element="time"
                        filter={[Function]}
                        format="DD-MM-YYYY"
                        fromNow={false}
                        interval={60000}
                        key="2"
                        onChange={[Function]}
                        toNow={false}
                        unit={null}
                        unix={false}
                        utc={false}
                      >
                        <time
                          dateTime={"2018-08-17T09:15:45.000Z"}
                        >
                          17-08-2018
                        </time>
                      </t>
                    </td>
                    <td
                      key="4"
                    >
                      <t
                        ago={false}
                        calendar={false}
                        decimal={false}
                        element="time"
                        filter={[Function]}
                        format="DD-MM-YYYY"
                        fromNow={false}
                        interval={60000}
                        key="2"
                        onChange={[Function]}
                        toNow={false}
                        unit={null}
                        unix={false}
                        utc={false}
                      >
                        <time
                          dateTime={null}
                        >
                          Invalid date
                        </time>
                      </t>
                    </td>
                    <td
                      key="5"
                    >
                      Form 2
                    </td>
                  </ElementMap>
                </tr>
              </tbody>
            </table>
          </Table>
          <Pagination
            aria-label="Page navigation example"
            listTag="ul"
            tag="nav"
          >
            <nav
              aria-label="Page navigation example"
              className=""
            >
              <ul
                className="pagination"
              >
                <PaginationItem
                  disabled={false}
                  tag="li"
                >
                  <li
                    className="page-item"
                  >
                    <Link
                      aria-label="First"
                      className="page-link"
                      replace={false}
                      to="/tasks/?page=first"
                    >
                      <a
                        aria-label="First"
                        className="page-link"
                        href="/tasks/?page=first"
                        onClick={[Function]}
                      >
                        First
                      </a>
                    </Link>
                  </li>
                </PaginationItem>
                <PaginationItem
                  disabled={true}
                  tag="li"
                >
                  <li
                    className="page-item disabled"
                  >
                    <Link
                      aria-label="Previous"
                      className="page-link"
                      replace={false}
                      to="#"
                    >
                      <a
                        aria-label="Previous"
                        className="page-link"
                        href="/"
                        onClick={[Function]}
                      >
                        «
                      </a>
                    </Link>
                  </li>
                </PaginationItem>
                <PaginationItem
                  disabled={true}
                  tag="li"
                >
                  <li
                    className="page-item disabled"
                  >
                    <Link
                      aria-label="Next"
                      className="page-link"
                      replace={false}
                      to="#"
                    >
                      <a
                        aria-label="Next"
                        className="page-link"
                        href="/"
                        onClick={[Function]}
                      >
                        »
                      </a>
                    </Link>
                  </li>
                </PaginationItem>
                <PaginationItem
                  disabled={false}
                  tag="li"
                >
                  <li
                    className="page-item"
                  >
                    <Link
                      aria-label="Last"
                      className="page-link"
                      replace={false}
                      to="/tasks/?page=last"
                    >
                      <a
                        aria-label="Last"
                        className="page-link"
                        href="/tasks/?page=last"
                        onClick={[Function]}
                      >
                        Last
                      </a>
                    </Link>
                  </li>
                </PaginationItem>
              </ul>
            </nav>
          </Pagination>
        </div>
      </ListView>
    </div>
  </TasksList>
</Router>
`;
